aws/resource_aws_sagemaker_notebook_instance.go

```diff
@@ -0,0 +1,371 @@
+		return err
+	}
+	sgs := notebookInstance.SecurityGroups
+	if sgs != nil && len(sgs) > 0 {
+		if err := d.Set("security_groups", fromStringPSliceToStringSliceSagemaker(&sgs)); err != nil {
+			log.Printf("[ERR] Error setting Security Groups: %s", err)
+			return err
+		}
+	}
+	if err := d.Set("kms_key_id", notebookInstance.KmsKeyId); err != nil {
+		log.Printf("[ERR] Error setting KMS Key ID: %s", err)
+		return err
+	}
+	if err := d.Set("creation_time", notebookInstance.CreationTime.Format(time.RFC3339)); err != nil {
+		log.Printf("[ERR] Error setting Creation Time: %s", err)
+		return err
+	}
+	if err := d.Set("last_modified_time", notebookInstance.LastModifiedTime.Format(time.RFC3339)); err != nil {
+		log.Printf("[ERR] Error setting Last Modified Time: %s", err)
+		return err
+	}
+	if err := d.Set("status", notebookInstance.NotebookInstanceStatus); err != nil {
+		log.Printf("[ERR] Error setting Status: %s", err)
+		return err
+	}
+	if err := d.Set("arn", notebookInstance.NotebookInstanceArn); err != nil {
+		log.Printf("[ERR] Error setting ARN: %s", err)
+		return err
+	}
+
+	tagsOutput, err := conn.ListTags(&sagemaker.ListTagsInput{
+		ResourceArn: notebookInstance.NotebookInstanceArn,
+	})
+	if err != nil {
+		log.Printf("[ERR] Error reading tags: %s", err)
+		return err
+	}
+
+	d.Set("tags", tagsToMapSagemaker(tagsOutput.Tags))
+
+	return nil
+}
+
+func resourceAwsSagemakerNotebookInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	d.Partial(true)
+
+	if err := setSagemakerTags(conn, d); err != nil {
+		return err
+	} else {
+		d.SetPartial("tags")
+	}
+
+	// Once a nootebook is created only the instance type and role can be modified
+	if d.HasChange("subnet_id") || d.HasChange("security_groups") || d.HasChange("kms_key_id") {
+		return fmt.Errorf("Error existing Notebook Instance cannot be updated")
+	}
+
+	hasChanged := false
+	// Update
+	updateOpts := &sagemaker.UpdateNotebookInstanceInput{
+		NotebookInstanceName: aws.String(d.Get("name").(string)),
+	}
+
+	if d.HasChange("role_arn") {
+		updateOpts.RoleArn = aws.String(d.Get("role_arn").(string))
+		hasChanged = true
+	}
+
+	if d.HasChange("instance_type") {
+		updateOpts.InstanceType = aws.String(d.Get("instance_type").(string))
+		hasChanged = true
+	}
+
+	if hasChanged {
+		// Stop notebook
+		_, previousStatus, _ := SagemakerNotebookInstanceStateRefreshFunc(conn, d.Id())()
+		if err := stopSagemakerNotebookInstance(conn, d, true); err != nil {
+			return fmt.Errorf("Error stopping Sagemaker Notebook Instance: %s", err)
+		}
+
+		if _, err := conn.UpdateNotebookInstance(updateOpts); err != nil {
+			return fmt.Errorf("Error updating Sagemaker Notebook Instance: %s", err)
+		}
+
+		// Restart if needed
+		if previousStatus == "InService" {
+			startOpts := &sagemaker.StartNotebookInstanceInput{
+				NotebookInstanceName: aws.String(d.Id()),
+			}
+
+			if _, err := conn.StartNotebookInstance(startOpts); err != nil {
```

Comment:
We should return the errors here if we can't get it back into the state it was previously